### ISSUE TITLE
SQL Bind/Cursor Issue

### ISSUE THREAD

**Issue by slewis30328** [oracle/node-oracledb#1694]
We have encountered an issue related to max open cursors. We have mapped the error back to a specific query. The query is a little different - binds to a collection type. Unfortunately, we are unable to replicate this. The problem has only surfaced 2X in the last month - obviously volume/load related. The internal references to Oracle packages is interesting. Not familiar with that package.  We recently upgraded from Oracle 19.12 to 19.24. I'm not 100% sure of this, but I don't think we ran into this issue prior to the upgrade.  Any ideas would be much appreciated.  

**NOTE:** :p_catMasterIDArr would never have more than 100 elements

**Query:**
SELECT CATMASTERID 
FROM   SMSYS.ITEMFAV
WHERE  LOCATIONID = :p_locID
AND    CATMASTERID IN (SELECT * FROM TABLE(:p_catMasterIDArr))

**Oracle Error:**
09/19/2024 12:03:29 ORA-01000: maximum open cursors exceeded
ORA-06512: at "SYS.DBMS_PICKLER", line 144
ORA-06512: at "SYS.DBMS_PICKLER", line 189
ORA-06512: at "SYS.DBMS_PICKLER", line 405
ORA-06512: at "SYS.DBMS_PICKLER", line 144

**Application Env:**
oracledb version: 6.5.1
Node version: node-v20.11.0-win-x64
Oracle DB version: 19.24.0.0.0
OS: Windows Server 2016

**Comment by slewis30328** [comment]
Just wanted to add an update to this issue.  After further testing, I can see that the cursor is never released/reused on this query. Every time the query runs, a new row is created in v$open_cursor. All my other queries are parsed and loaded 1X and I do not see any additions to v$open_cursor.  I think this issue should be flagged as a bug at this time.

**Comment by sudarshan12s** [comment]
Thanks @slewis30328  for the details, we will try to reproduce it . Can you give more details :
-  Is the query run from pl/sql block? I mean like below or its just a sql query. any minimal program would help.  
-  I believe its run in thin mode not in thick mode (`oracledb.initOracleClient `) 

```

const tableName = 'itemfav';
const typename = 'catmasterid_arr';

const sql = `          
-- Create a table similar to SMSYS.ITEMFAV
CREATE TABLE if not exists ${tableName} (
    catmasterid NUMBER,
    locationid NUMBER
)`;
    await connection.execute(sql);

await connection.execute(`-- PL/SQL anonymous block with collection bind
                    DECLARE
                        p_catMasterIDArr ${typename} := ${typename}();
                        p_locID NUMBER := 100;
                        CURSOR c1 IS
                            SELECT catmasterid
                            FROM ${tableName}
                            WHERE locationid = p_locID
                            AND catmasterid IN (SELECT * FROM TABLE(p_catMasterIDArr));
                        v_catMasterID NUMBER;
                    BEGIN
                        -- Populate the collection with some test data
                        FOR i IN 1..100 LOOP
                            p_catMasterIDArr.EXTEND;
                            p_catMasterIDArr(i) := i;
                        END LOOP;
                    
                        -- Open the cursor and fetch results
                        OPEN c1;
                        LOOP
                            FETCH c1 INTO v_catMasterID;
                            EXIT WHEN c1%NOTFOUND;
                            DBMS_OUTPUT.PUT_LINE('CATMASTERID: ' || v_catMasterID);
                        END LOOP;
                    
                        -- Close the cursor
                        CLOSE c1;
                    END;`);
```

**Comment by slewis30328** [comment]
It's just run as a sql query and we are using thin mode.

Here is some DDL examples if needed:

CREATE OR REPLACE type SMSYS.NumberArray as table of number

CREATE TABLE SMSYS.ITEMFAV
(
  ITEMFAVID     NUMBER                          NOT NULL,
  LOCATIONID    NUMBER                          NOT NULL,
  CATMASTERID   NUMBER                          NOT NULL,
  ITEMFAVCATID  NUMBER                          NOT NULL,
  CREATEDATE    DATE                            NOT NULL,
  CREATEUSER    VARCHAR2(50 BYTE)               NOT NULL,
  UPDATEDATE    DATE,
  UPDATEUSER    VARCHAR2(50 BYTE)
)

NOTE:  I didn't add index DDL - let me know if you need that.

**Comment by sudarshan12s** [comment]
I am trying to add a test as below to reproduce the issue seen , can you check and let me know if it represents the issue at high level.

```

'use strict';

const oracledb = require('oracledb');
const dbConfig = require('./dbconfig.js');
const testsUtil = require('../test/testsUtil.js');

async function run() {

  let connection, sysDBAConn;
  const tableName = 'itemfav';
  const typename = 'CATMASTERIDARR'; // keep in capital

  try {
    connection = await oracledb.getConnection(dbConfig);

    const dbaConfig = {
      user: dbConfig.DBA_user,
      password: dbConfig.DBA_password,
      connectionString: dbConfig.connectString,
      privilege: oracledb.SYSDBA
    };
    sysDBAConn = await oracledb.getConnection(dbaConfig);
    const sid = await testsUtil.getSid(connection);

    let openCount = await testsUtil.getOpenCursorCount(sysDBAConn, sid);
    console.log(' open cursor count to start ', openCount);

    let sql = `
-- Create a table similar to SMSYS.ITEMFAV
CREATE TABLE  ${tableName} (
    catmasterid NUMBER,
    locationid NUMBER
)`;
    await connection.execute(sql);

    const binds = [];
    for (let index = 0; index < 100; index++) {
      binds.push([index + 1, 100]);
    }
    await connection.executeMany(`INSERT INTO ${tableName} (catmasterid, locationid) VALUES (:1, :2)`, binds);

    sql = ` -- Create a PL/SQL collection type
                CREATE OR REPLACE TYPE ${typename} IS TABLE OF NUMBER;`;
    await testsUtil.createType(connection, typename, sql);

    const objData = Array.from({ length: 100 }, (_, index) => index + 1);
    const objClass = await connection.getDbObjectClass(typename);
    const testObj = new objClass(objData);

    // try increased load by looping 
    const iterations = 10000;  // (increase as needed)

    for (let i = 1; i <= iterations; i++) {
      const result = await connection.execute(`SELECT catmasterid
    FROM ${tableName}
    WHERE locationid = 100
    AND catmasterid IN (SELECT * FROM TABLE(:p_catMasterIDArr))`, [testObj]);
      //console.log(result);
    }

    openCount = await testsUtil.getOpenCursorCount(sysDBAConn, sid);
    console.log(' open count after  all iterations ', openCount);
  } catch (err) {
    console.log(err);
  } finally {
    await connection.execute(`TRUNCATE TABLE ${tableName}`);
    await connection.close();
    await sysDBAConn.close();
  }
}

run(); 
```

**Comment by sudarshan12s** [comment]
@slewis30328 , I could not reproduce the issue with 19.24 DB version. Can you let us know if above program reproduces the issue in your environment or if the above test needs modifications to reproduce.

**Comment by slewis30328** [comment]
When I get a chance, I will see if I can reproduce issue with your test case above.  Not sure if this makes a difference, but we are using connection pooling - example above is not using that. When I run my tests, this is the SQL statement that gets repeated and causes the open cursor count to increase for each execution.  I guess this statement is run from your driver layer?

SELECT /*+ NOPARALLEL */
         1,
         A.NAME,
         A.ATTRIBUTE#,
         DECODE (AT.TYPECODE,
                 9, DECODE (A.CHARSETFORM, 2, 'NVARCHAR2', ATO.NAME),
                 96, DECODE (A.CHARSETFORM, 2, 'NCHAR', ATO.NAME),
                 112, DECODE (A.CHARSETFORM, 2, 'NCLOB', ATO.NAME),
                 ATO.NAME),
         DECODE (BITAND (AT.PROPERTIES, 64), 64, NULL, ATU.NAME),
         NULL,
         A.ATTR_TOID,
         DECODE (BITAND (T.PROPERTIES, 65536), 65536, 'NO', 'YES'),
         SU.NAME,
         SO.NAME
    FROM SYS.ATTRIBUTE$            A,
         SYS.TYPE$                 T,
         SYS.TYPE$                 AT,
         SYS."_CURRENT_EDITION_OBJ" ATO,
         SYS.USER$                 ATU,
         SYS."_CURRENT_EDITION_OBJ" SO,
         SYS.USER$                 SU
   WHERE     T.TVOID = :B1
         AND A.ATTR_TOID = ATO.OID$
         AND ATO.OWNER# = ATU.USER#
         AND A.TOID = T.TVOID
         AND T.PACKAGE_OBJ# IS NULL
         AND AT.TVOID = A.ATTR_TOID
         AND AT.SUPERTOID = SO.OID$(+)
         AND SO.OWNER# = SU.USER#(+)
ORDER BY ATTRIBUTE#

**Comment by sudarshan12s** [comment]
> When I get a chance, I will see if I can reproduce issue with your test case above. Not sure if this makes a difference, but we are using connection pooling - example above is not using that. When I run my tests, this is the SQL statement that gets repeated and causes the open cursor count to increase for each execution. I guess this statement is run from your driver layer?
> 
> SELECT /*+ NOPARALLEL */ 1, A.NAME, A.ATTRIBUTE#, DECODE (AT.TYPECODE, 9, DECODE (A.CHARSETFORM, 2, 'NVARCHAR2', ATO.NAME), 96, DECODE (A.CHARSETFORM, 2, 'NCHAR', ATO.NAME), 112, DECODE (A.CHARSETFORM, 2, 'NCLOB', ATO.NAME), ATO.NAME), DECODE (BITAND (AT.PROPERTIES, 64), 64, NULL, ATU.NAME), NULL, A.ATTR_TOID, DECODE (BITAND (T.PROPERTIES, 65536), 65536, 'NO', 'YES'), SU.NAME, SO.NAME FROM SYS.ATTRIBUTE$ A, SYS.TYPE$ T, SYS.TYPE$ AT, SYS."_CURRENT_EDITION_OBJ" ATO, SYS.USER$ ATU, SYS."_CURRENT_EDITION_OBJ" SO, SYS.USER$ SU WHERE T.TVOID = :B1 AND A.ATTR_TOID = ATO.OID$ AND ATO.OWNER# = ATU.USER# AND A.TOID = T.TVOID AND T.PACKAGE_OBJ# IS NULL AND AT.TVOID = A.ATTR_TOID AND AT.SUPERTOID = SO.OID$(+) AND SO.OWNER# = SU.USER#(+) ORDER BY ATTRIBUTE#

Thanks. I tried with pool. I modified my test using pool. I am able to see the issue, I will share more details.  
```

for (let i = 1; i <= iterations; i++) {
      connection = await pool.getConnection();
      const objClass = await connection.getDbObjectClass(typename);
      const testObj = new objClass(objData);
      const result = await connection.execute(`SELECT catmasterid
    FROM ${tableName}
    WHERE locationid = 100
    AND catmasterid IN (SELECT * FROM TABLE(:p_catMasterIDArr))`, [testObj]);
      await connection.close();
    }
```

**Comment by slewis30328** [comment]
Can I apply this to 6.6.0 or is this just for 6.5.1?  My dev env is 6.6.0 and we have 6.5.1 in TEST and PROD.

**Comment by slewis30328** [comment]
I tested with 6.5.1 and that fix above worked.  Do not see the cursor count increase.

**Comment by sudarshan12s** [comment]
I have done more testing. I am sharing the patch for 6.5.1 . The issue was happening when the `getDbObjectClass`  was done on a given `type` multiple times between connection acquire and release to pool. 
Yes it is safe to be applied in production. This fix will be available in the next release. Thanks.

```diff
diff --git a/lib/thin/connection.js b/lib/thin/connection.js
index 7a8a90548..db9ca7904 100644
--- a/lib/thin/connection.js
+++ b/lib/thin/connection.js
@@ -513,31 +513,37 @@ class ThinConnectionImpl extends ConnectionImpl {
     // check cache; if already present, nothing more to do!
     const info = this._getDbObjectType(result.outBinds.schema,
       result.outBinds.name, result.outBinds.package_name, result.outBinds.oid);
-    if (!info.partial)
+    if (!info.partial) {
+      result.outBinds.attrs_rc.close();
       return info;
+    }

+    try {
     // process TDS and attributes cursor
-    info.version = result.outBinds.version;
-    const attrRows = await result.outBinds.attrs_rc.getRows(1000, {});
-    if (attrRows.length > 0) {
+      info.version = result.outBinds.version;
+      const attrRows = await result.outBinds.attrs_rc.getRows(1000, {});
+      if (attrRows.length > 0) {
       // Its an object not a collection.
-      info.attributes = [];
-      for (const row of attrRows) {
-        const attr = { name: row[1] };
-        if (row[4]) {
-          attr.type = types.DB_TYPE_OBJECT;
-          attr.typeClass = this._getDbObjectType(row[4], row[3], row[5], row[6]);
-          if (attr.typeClass.partial) {
-            this._partialDbObjectTypes.push(attr.typeClass);
+        info.attributes = [];
+        for (const row of attrRows) {
+          const attr = { name: row[1] };
+          if (row[4]) {
+            attr.type = types.DB_TYPE_OBJECT;
+            attr.typeClass = this._getDbObjectType(row[4], row[3], row[5], row[6]);
+            if (attr.typeClass.partial) {
+              this._partialDbObjectTypes.push(attr.typeClass);
+            }
+          } else {
+            attr.type = types.getTypeByColumnTypeName(row[3]);
           }
-        } else {
-          attr.type = types.getTypeByColumnTypeName(row[3]);
+          info.attributes.push(attr);
         }
-        info.attributes.push(attr);
       }
-    }

-    await this._parseTDS(result.outBinds.tds, info);
+      await this._parseTDS(result.outBinds.tds, info);
+    } finally {
+      result.outBinds.attrs_rc.close();
+    }
     info.partial = false;
     return info;
```



**Note:** I have added finally block to close in all cases. Without tabs (`git diff -w`) , the diff is minor

**Comment by slewis30328** [comment]
Is it possible to post the updated connection.js file from your patch? Thank you.

**Comment by sudarshan12s** [comment]
> Is it possible to post the updated connection.js file from your patch? Thank you.

[connection.txt](https://github.com/user-attachments/files/17177665/connection.txt)

**Comment by ichursin** [comment]
Hi, looks like we are still able to reproduce the same behavior on 6.6.0 version

We've use static pool size (minPool equals maxPool) that means our connections are always present

And... our connections after random period of time reporting such kind of error: ORA-04025: maximum allowed library object lock allocated for stringstringstring

Deep investigation confirm that query below has state OPEN and it's increases while application becomes unavailable

```
SELECT /*+ NOPARALLEL */
1,
A.NAME,
A.ATTRIBUTE#,
DECODE (AT.TYPECODE,
9, DECODE (A.CHARSETFORM, 2, 'NVARCHAR2', ATO.NAME),
96, DECODE (A.CHARSETFORM, 2, 'NCHAR', ATO.NAME),
112, DECODE (A.CHARSETFORM, 2, 'NCLOB', ATO.NAME),
ATO.NAME),
DECODE (BITAND (AT.PROPERTIES, 64), 64, NULL, ATU.NAME),
NULL,
A.ATTR_TOID,
DECODE (BITAND (T.PROPERTIES, 65536), 65536, 'NO', 'YES'),
SU.NAME,
SO.NAME
FROM SYS.ATTRIBUTE$ A,
SYS.TYPE$ T,
SYS.TYPE$ AT,
SYS."_CURRENT_EDITION_OBJ" ATO,
SYS.USER$ ATU,
SYS."_CURRENT_EDITION_OBJ" SO,
SYS.USER$ SU
WHERE T.TVOID = :B1
AND A.ATTR_TOID = ATO.OID$
AND ATO.OWNER# = ATU.USER#
AND A.TOID = T.TVOID
AND T.PACKAGE_OBJ# IS NULL
AND AT.TVOID = A.ATTR_TOID
AND AT.SUPERTOID = SO.OID$(+)
AND SO.OWNER# = SU.USER#(+)
ORDER BY ATTRIBUTE#
```

Do you have any ideas regarding that? :)

**Comment by sudarshan12s** [comment]
Is it possible to log these details on open cursor details. The complete program is [here](https://github.com/oracle/node-oracledb/issues/1694#issuecomment-2372994520).


```
result = await sysDBAConn.execute(`SELECT
    SID,
    USER_NAME,
    ADDRESS,
    HASH_VALUE,
    SQL_ID,
    SQL_TEXT,
    CURSOR_TYPE
FROM
    V$OPEN_CURSOR
WHERE
    SID = ${sid}`);
    console.log(`details of open cursors `, result.rows);
```

As i understand the issue , ORA-01000 error is seen even after [patch](https://github.com/oracle/node-oracledb/issues/1694#issuecomment-2377536816) is applied on top of node-oracledb 6.6.0.

**Comment by ichursin** [comment]
@sharadraju hm.... looks like patch from 6.5.1 could not be applied correctly to 6.6.0 cause it's broken by this commit https://github.com/oracle/node-oracledb/commit/c7b3c31d5110e478ed602c99308ead41d81f83d7

**Comment by sudarshan12s** [comment]
sorry small correction: Yes the patch shared [above](https://github.com/oracle/node-oracledb/issues/1694#issuecomment-2377536816) is on top of 6.5.1

**Comment by sharadraju** [comment]
@ichursin The patch shared above has to be applied on top of 6.5.1, not on 6.6 as it has other changes

We will be releasing 6.7 in a few days anyway, which will have the complete fix.

**Comment by ichursin** [comment]
@sharadraju could you share changes regarding 6.7 version?

cause it's affect our production systems today :)

**Comment by sharadraju** [comment]
@ichursin Note that the patch was tested on top of 6.5.1, not on 6.6. We generally release the patches to the customers to give a quick fix for their issue. If there is no urgency, it is better to wait for the next release, given that it will be out very soon ;)

As for the fix, the change is a slightly big one. We will try to see if we can send the latest version of the patched file(s). This patched fileset has been tested with the upcoming 6.7 version, not the older versions. So the risk of using it lies with you, the user.

Another alternative can be to downgrade to 6.5.1 and add the patch on top of it. It should work, if it is urgent.

As I said, 6.7 is due very soon (sooner than you expect!) :). Please wait for it, if possible.

**Comment by sharadraju** [comment]
@ichursin Are you ok waiting for the 6.7 release?

**Comment by ichursin** [comment]
yep, we've going to downgrade 6.5.1 and wait 6.7

**Comment by sharadraju** [comment]
This is fixed in 6.7 release.
@ichursin FYI
